crates/ruff/tests/integration_test.rs

```diff
@@ -1523,7 +1581,17 @@ fn check_hints_hidden_unsafe_fixes() {
     exit_code: 1
     ----- stdout -----
     -:1:1: RUF901 [*] Hey this is a stable test rule with a safe fix.
+    Suggested fix:
+      1 |+# fix from stable-test-rule-safe-fix
+
+        Run `ruff check --fix` to apply this fix.
+
     -:1:1: RUF902 Hey this is a stable test rule with an unsafe fix.
+    Suggested fix:
+      1 |+# fix from stable-test-rule-unsafe-fix
+
+        Run `ruff check --fix --unsafe-fixes` to apply this unsafe fix.
```

Comment:
Given the message below this ("1 hidden fix..."), I wonder if we should be hiding unsafe fixes unless the opt-in is provided. I think this would solve the "how do we display applicability" problem, to an extent?

If so, I think we could figure out what to do with "display-only" fixes later and just _never_ show them to start (to limit the scope of this pull request)

Reply:
This would require a bit of futzing to retain the existing coverage in our test suite, maybe just using `--unsafe-fixes` by default in some places?

Reply:
Hmm, yeah—I can definitely back out those parts of the change if that’s preferable!

Reply:
It feels like the current design (prior to this pull request) _hides_ unsafe fixes and allowing _opt-in_ to their display, at which point we don't need to indicate that they're unsafe? It seems easiest to follow that pattern for now and consider if and how we want to display unsafe fixes by default separately, i.e., I think more discussion in an issue would be prudent.

Reply:
In terms of handling applicability, we could do that in a couple of ways:

- The `Diff` itself could take a configuration level and its `Display` implementation could use that to decide what to print.
- The caller could decide whether to print based on message applicability at a higher level, keeping it out of the `Diff`. (This is vaguely how I would lean, personally.

The latter is a slightly smaller scope and IMO better isolation of concerns; it keeps it properly in the `TextEmitter`, whereas the `Diff` can be concerned entirely with the presentation of a given diff.

If we go that way, we still have the choice whether to keep the applicability message in the diff. I think I am still mildly inclined toward showing that CTA, because of the differences when you *do* include `--unsafe-fixes`; it seems like it would still be nice to know which changes are which.

Reply:
It seems reasonable to decide whether or not the `Diff` should be displayed outside its display implementation (the latter option).

Regarding the CTA, I have a preference for splitting it out of this pull request so we can merge this sooner and have focused discussion on each objective, like:

1. Display available fixes by default
2. Disambiguate safe and unsafe fixes when displayed (i.e., via a CTA or other label)
3. Display unsafe fixes by default (needs discussion)
4. Determine how to make display-only fixes user-facing

What do you think?

Reply:
That makes good sense to me. I will split out issues for further discussion on those points and update this to simply display fixes only for the `Applicability::Safe` level, in a slightly clunky way that is amenable to cleaning up by way of broader refactors once some of those other design questions are fleshed out.